### DISCUSSION TITLE
EN: Driver for USB Keyboard / ES: Driver para Teclado USB

### DISCUSSION THREAD

**Discussion by ghost** [CosmosOS/Cosmos#2669] [Q&A]
English:
Hello Community of CosmosOS,

Can someone help me to create a Driver for USB Keyboard.
ChatGPT gave me a code but it contains errors, the code is at the end of this Post.
Can someone help me please.

Español:
Hola Comunidad de CosmosOS,

Alguien puede ayudarme a crear un Driver para Teclado USB.
ChatGPT me entregó un código pero este contiene errores, el código está al final de esta Publicación.
Alguien puede ayudarme por favor.

> using System;
> using System.Linq;
> using HidLibrary;
> 
> namespace StellarOS
> {
>     class Kernel
>     {
>         static void Main(string[] args)
>         {
>             
>             var keyboardFilter = new HidDeviceLoader().GetDeviceFilter(vendorID: 0x1234, productID: 0x5678); // Reemplaza con los valores correctos
> 
>             var keyboardDevices = new HidDeviceLoader().GetDevices(keyboardFilter);
>             var keyboardDevice = keyboardDevices.FirstOrDefault();
>             if (keyboardDevice != null)
>             {
>                 Console.WriteLine("Keyboard Connected");
> 
>                 keyboardDevice.OpenDevice();
> 
>                 keyboardDevice.Inserted += KeyboardInserted;
>                 keyboardDevice.Removed += KeyboardRemoved;
>                 keyboardDevice.MonitorDeviceEvents = true;
>                 keyboardDevice.ReadReport(OnReport);
> 
>                 Console.WriteLine("Press any key to exit...");
>                 Console.ReadKey();
> 
>                 keyboardDevice.CloseDevice();
>             }
>             else
>             {
>                 Console.WriteLine("Keyboard Not Found");
>             }
>         }
> 
>         static void KeyboardInserted()
>         {
>             Console.WriteLine("Keyboard Conected");
>         }
> 
>         static void KeyboardRemoved()
>         {
>             Console.WriteLine("Keyboard Disconected");
>         }
> 
>         static void OnReport(HidReport report)
>         {
>             
>         }
>     }
> }

**Comment by retrievaaaa** [comment]
1) NEVER use ChatGPT on Cosmos, it barely knows what it is and will hallucinate A LOT
2) this isnt a usb driver, this looks like its using some sort of library which definitely does not work in cosmos

You have two options:
- Implement USB yourself
- Use the already existing PS2 keyboard driver

**Reply by ghost** [reply]
Hello GoldenretriverYT,

Thank you for answering but I asked how can I create a driver, 
I also appreciate your prompt response

**Reply by retrievaaaa** [reply]
This a way too broad topic for us to cover here. Please do research on your own.

**Comment by zarlo** [comment]
https://wiki.osdev.org/Universal_Serial_Bus

would be a good starting point
